File: src/server/src/test/test_velvet_dawn/test_turns.py
import time

from velvet_dawn import errors
import velvet_dawn.players
from test.base_test import BaseTest
from velvet_dawn.config import Config
from velvet_dawn.server.app import config
from velvet_dawn.dao import app, db
from velvet_dawn.dao.models import Keys, Player
from velvet_dawn.models.phase import Phase


class TestGameTurns(BaseTest):

    def test_check_end_turn_case_in_setup_from_time(self):
        """ Test if the turn should end """
        with app.app_context():
            test_config = Config()
            test_config.setup_time = 1

            velvet_dawn.players.join("player1", "password")
            velvet_dawn.players.join("player2", "password")
            velvet_dawn.game.phase.start_setup_phase(test_config)

            time.sleep(2)
            velvet_dawn.game.turns.check_end_turn_case(test_config)
            self.assertEqual(velvet_dawn.game.phase.get_phase(), Phase.GAME)

    def test_check_end_turn_case_in_setup_when_ready(self):
        """ Test if the turn should end """
        with app.app_context():
            velvet_dawn.players.join("player1", "password")
            velvet_dawn.players.join("player2", "password")
            velvet_dawn.game.phase.start_setup_phase(config)

            db.session.query(Player).update({Player.ready: True})
            db.session.commit()
            velvet_dawn.game.turns.check_end_turn_case(config)
            self.assertEqual(velvet_dawn.game.phase.get_phase(), Phase.GAME)

    def test_check_end_turn_case_in_game_from_time(self):
        """ Test if the turn should end """
        with app.app_context():
            test_config = Config()
            test_config.turn_time = 1

            velvet_dawn.players.join("player1", "password")
            velvet_dawn.players.join("player2", "password")
            velvet_dawn.game.phase.start_game_phase()
            velvet_dawn.game.turns._update_turn_start_time()
            self.assertEqual(
                velvet_dawn.players.get_player("player1").team,
                velvet_dawn.game.turns.get_active_turn(Phase.GAME)
            )

            time.sleep(2)
            velvet_dawn.game.turns.check_end_turn_case(test_config)
            self.assertEqual(
                velvet_dawn.players.get_player("player2").team,
                velvet_dawn.game.turns.get_active_turn(Phase.GAME)
            )

    def test_check_end_turn_case_in_game_when_ready(self):
        """ Test if the turn should end """
        with app.app_context():
            velvet_dawn.players.join("player1", "password")
            velvet_dawn.players.join("player2", "password")
            velvet_dawn.game.phase.start_game_phase()

            self.assertEqual(
                velvet_dawn.players.get_player("player1").team,
                velvet_dawn.game.turns.get_active_turn(Phase.GAME)
            )

            db.session.query(Player).where(Player.name == "player1").update({Player.ready: True})
            db.session.commit()
            velvet_dawn.game.turns.check_end_turn_case(config)
            self.assertEqual(
                velvet_dawn.players.get_player("player2").team,
                velvet_dawn.game.turns.get_active_turn(Phase.GAME)
            )

    def test_begin_next_turn_resets_entity_movement(self):
        """ Test that when a turn begins the entity movement is set to it's range """
        with app.app_context():
            test_config = Config().load()
            test_config.seed = 0

            # Setup the game state
            velvet_dawn.datapacks.init(test_config)
            velvet_dawn.map.new(test_config)
            velvet_dawn.players.join("player1", "password")
            velvet_dawn.game.phase._set_phase(Phase.Lobby)
            velvet_dawn.game.setup.update_setup("civil-war:commander", 1)
            velvet_dawn.game.phase.start_setup_phase(test_config)
            velvet_dawn.game.setup.place_entity("player1", "civil-war:commander", test_config.map_width // 2, 0)

            unit = velvet_dawn.units.list("player1")[0]
            self.assertEqual(0, unit.get_attribute("movement.remaining", _type=int))

            velvet_dawn.game.phase.start_game_phase()

            self.assertEqual(
                unit.get_attribute("movement.remaining", _type=int),
                unit.get_attribute("movement.range", _type=int)
            )

    def test_begin_next_turn_and_get_active_turn(self):
        """ This will test getting the active turn and beginning a new turn """
        with app.app_context():
            velvet_dawn.players.join("player1", "password")
            velvet_dawn.players.join("player2", "password")
            self.assertIsNone(velvet_dawn.game.turns.get_active_turn(Phase.Setup))

            velvet_dawn.game.phase.start_game_phase()

            self.assertEqual(
                velvet_dawn.players.get_player("player1").team,
                velvet_dawn.game.turns.get_active_turn(Phase.GAME)
            )

            velvet_dawn.game.turns.ready("player1")

            velvet_dawn.game.turns.begin_next_turn()

            players = velvet_dawn.players.list()
            for player in players:
                self.assertFalse(player.ready)

            self.assertEqual(
                velvet_dawn.players.get_player("player2").team,
                velvet_dawn.game.turns.get_active_turn(Phase.GAME)
            )

            velvet_dawn.game.turns.begin_next_turn()

            self.assertEqual(
                velvet_dawn.players.get_player("player1").team,
                velvet_dawn.game.turns.get_active_turn(Phase.GAME)
            )

    def test_check_all_players_ready_during_setup(self):
        with app.app_context():
            velvet_dawn.players.join("player1", "password")
            velvet_dawn.players.join("player2", "password")
            velvet_dawn.players.join("player3", "password")

            velvet_dawn.game.turns.ready("player1")
            self.assertFalse(velvet_dawn.game.turns._check_all_players_ready(Phase.Setup))
            velvet_dawn.game.turns.ready("player2")
            self.assertFalse(velvet_dawn.game.turns._check_all_players_ready(Phase.Setup))
            velvet_dawn.game.turns.ready("player3")
            self.assertTrue(velvet_dawn.game.turns._check_all_players_ready(Phase.Setup))

            velvet_dawn.game.phase.start_game_phase()

            velvet_dawn.game.turns.ready("player2")
            self.assertFalse(velvet_dawn.game.turns._check_all_players_ready(Phase.GAME))
            velvet_dawn.game.turns.ready("player3")
            self.assertFalse(velvet_dawn.game.turns._check_all_players_ready(Phase.GAME))
            velvet_dawn.game.turns.ready("player1")
            self.assertTrue(velvet_dawn.game.turns._check_all_players_ready(Phase.GAME))

    def test_ready_setup_placed_commander(self):
        with app.app_context():
            velvet_dawn.players.join("player1", "password")
            velvet_dawn.game.phase._set_phase(Phase.Setup)

            # No commanders played so can't ready up
            with self.assertRaises(errors.ValidationError):
                velvet_dawn.game.turns.ready("player1")

    def test_ready_unready(self):
        with app.app_context():
            velvet_dawn.players.join("player1", "password")
            velvet_dawn.players.join("player2", "password")

            velvet_dawn.game.turns.ready("player1")
            self.assertTrue(velvet_dawn.players.get_player("player1").ready)
            self.assertFalse(velvet_dawn.players.get_player("player2").ready)

            velvet_dawn.game.turns.ready("player2")
            self.assertTrue(velvet_dawn.players.get_player("player1").ready)
            self.assertTrue(velvet_dawn.players.get_player("player2").ready)

            velvet_dawn.game.turns.unready("player1")
            self.assertFalse(velvet_dawn.players.get_player("player1").ready)
            self.assertTrue(velvet_dawn.players.get_player("player2").ready)

            velvet_dawn.game.turns.unready("player2")
            self.assertFalse(velvet_dawn.players.get_player("player1").ready)
            self.assertFalse(velvet_dawn.players.get_player("player2").ready)

    def test_update_turn_start_time(self):
        with app.app_context():
            self.assertIsNone(velvet_dawn.dao.get_value(Keys.TURN_START))

            updated_time = velvet_dawn.game.turns._update_turn_start_time()
            self.assertEqual(round(updated_time), round(velvet_dawn.dao.get_value(Keys.TURN_START, _type=float)))

    def test_current_turn_time(self):
        with app.app_context():
            config = Config()

            config.turn_time = 10
            self.assertEqual(10, velvet_dawn.game.turns._current_turn_time(config, Phase.GAME))

            config.turn_time = 500
            self.assertEqual(500, velvet_dawn.game.turns._current_turn_time(config, Phase.GAME))

            # Test setup is always 5 mins
            config.setup_time = 5
            self.assertEqual(5, velvet_dawn.game.turns._current_turn_time(config, Phase.Setup))

File: src/jserver/src/test/velvetdawn/TestTurns.java
package test.velvetdawn;

import static org.junit.Assert.*;
import org.junit.Test;
import test.BaseTest;
import velvetdawn.VelvetDawn;
import velvetdawn.constants.AttributeKeys;
import velvetdawn.models.Coordinate;
import velvetdawn.models.Phase;
import velvetdawn.models.config.Config;

import java.util.ArrayList;

public class TestTurns extends BaseTest {

    /** Test if the turn should end */
    @Test
    public void test_check_end_turn_case_in_setup_from_time() throws Exception {
        var config = this.getConfig();
        config.setupTime = 1;
        var velvetDawn = new VelvetDawn(config);

        velvetDawn.players.join("player1", "password");
        velvetDawn.players.join("player2", "password");
        velvetDawn.game.startSetupPhase();

        assertEquals(velvetDawn.game.phase, Phase.Setup);
        Thread.sleep(2000);
        velvetDawn.game.turns.checkEndTurnCase();

        assertEquals(velvetDawn.game.phase, Phase.Game);
    }

    /** Test if the turn should end */
    @Test
    public void test_check_end_turn_case_in_setup_when_ready() throws Exception {
        var velvetDawn = new VelvetDawn(this.getConfig());

        var p1 = velvetDawn.players.join("player1", "password");
        var p2 = velvetDawn.players.join("player2", "password");
        velvetDawn.game.startSetupPhase();

        p1.ready = true;
        p2.ready = true;
        velvetDawn.game.turns.checkEndTurnCase();

        assertEquals(velvetDawn.game.phase, Phase.Game);
    }

    /** Test if the turn should end */
    @Test
    public void test_check_end_turn_case_in_game_from_time() throws Exception {
        var config = this.getConfig();
        config.turnTime = 1000;
        var velvetDawn = new VelvetDawn(config);

        var p1 = velvetDawn.players.join("player1", "password");
        var p2 = velvetDawn.players.join("player2", "password");
        velvetDawn.game.startGamePhase();
        velvetDawn.game.turns.updateTurnStartTime();
        assertEquals(p1.team, velvetDawn.game.turns.getActiveTurn());

        Thread.sleep(2000);
        velvetDawn.game.turns.checkEndTurnCase();
        assertEquals(p2.team, velvetDawn.game.turns.getActiveTurn());
    }

    /** Test if the turn should end */
    @Test
    public void test_check_end_turn_case_in_game_when_ready() throws Exception {
        var config = this.getConfig();
        var velvetDawn = new VelvetDawn(config);

        var p1 = velvetDawn.players.join("player1", "password");
        var p2 = velvetDawn.players.join("player2", "password");
        velvetDawn.game.startGamePhase();

        assertEquals(p1.team, velvetDawn.game.turns.getActiveTurn());

        p1.ready = true;
        velvetDawn.game.turns.checkEndTurnCase();
        assertEquals(p2.team, velvetDawn.game.turns.getActiveTurn());
    }

    /** Test that when a turn begins the entity movement is set to it's range */
    @Test
    public void test_begin_next_turn_resets_entity_movement() throws Exception {
        Config config = this.getConfig();
        VelvetDawn velvetDawn = new VelvetDawn(config);

        // Setup the game state
        var p1 = velvetDawn.players.join("player1", "null");
        velvetDawn.game.phase = Phase.Lobby;
        velvetDawn.game.setup.updateSetup("testing:commander", 1);
        velvetDawn.game.startSetupPhase();
        velvetDawn.game.setup.placeEntity(p1, "testing:commander", new Coordinate((float) (this.getConfig().map.width / 2), 0));

        var unit = new ArrayList<>(p1.entities).get(0);
        assertEquals(0, unit.attributes.get("movement.remaining").toNumber(), 0);

        velvetDawn.game.startGamePhase();

        assertEquals(
            unit.attributes.get(AttributeKeys.EntityRemainingMoves),
            unit.attributes.get("movement.range")
        );
    }

    /** This will test getting the active turn and beginning a new turn */
    @Test
    public void test_begin_next_turn_and_get_active_turn() throws Exception {
        var velvetDawn = new VelvetDawn(this.getConfig());

        var p1 = velvetDawn.players.join("player1", "password");
        var p2 = velvetDawn.players.join("player2", "password");
        assertNull(velvetDawn.game.turns.getActiveTurn());

        velvetDawn.game.startGamePhase();

        assertEquals(p1.team, velvetDawn.game.turns.getActiveTurn());

        velvetDawn.game.turns.ready(p1);

        velvetDawn.game.turns.beginNextTurn();

        assertFalse(p1.ready);
        assertFalse(p2.ready);

        assertEquals(p2.team, velvetDawn.game.turns.getActiveTurn());

        velvetDawn.game.turns.beginNextTurn();

        assertEquals(p1.team, velvetDawn.game.turns.getActiveTurn());
    }

    @Test
    public void test_check_all_players_ready_during_setup() throws Exception {
        Config config = this.getConfig();
        var velvetDawn = new VelvetDawn(config);

        velvetDawn.game.setup.updateSetup("testing:commander", 1);
        var p1 = velvetDawn.players.join("player1", "password");
        var p2 = velvetDawn.players.join("player2", "password");
        var p3 = velvetDawn.players.join("player3", "password");

        velvetDawn.game.startSetupPhase();
        velvetDawn.map.generate();

        velvetDawn.game.setup.placeEntity(p1, "testing:commander", new Coordinate(5, 0));
        velvetDawn.game.setup.placeEntity(p2, "testing:commander", new Coordinate(10, 6));
        velvetDawn.game.setup.placeEntity(p3, "testing:commander", new Coordinate(2, 10));

        velvetDawn.game.turns.ready(p1);
        assertFalse(velvetDawn.game.turns.checkAllPlayersReady());
        velvetDawn.game.turns.ready(p2);
        assertFalse(velvetDawn.game.turns.checkAllPlayersReady());
        velvetDawn.game.turns.ready(p3);
        assertTrue(velvetDawn.game.turns.checkAllPlayersReady());

        velvetDawn.game.startGamePhase();

        velvetDawn.game.turns.ready(p2);
        assertFalse(velvetDawn.game.turns.checkAllPlayersReady());
        velvetDawn.game.turns.ready(p3);
        assertFalse(velvetDawn.game.turns.checkAllPlayersReady());
        velvetDawn.game.turns.ready(p1);
        assertTrue(velvetDawn.game.turns.checkAllPlayersReady());
    }

    @Test
    public void test_ready_setup_placed_commander() throws Exception {
        var velvetDawn = new VelvetDawn(this.getConfig());

        var p1 = velvetDawn.players.getPlayer("player1");
        velvetDawn.game.phase = Phase.Setup;

        // No commanders played so can't ready up
        assertThrows(Exception.class, () ->
            velvetDawn.game.turns.ready(p1));
    }

    @Test
    public void test_ready_unready() throws Exception {
        var velvetDawn = this.prepareGame();
        var p1 = velvetDawn.players.getPlayer("player1");
        var p2 = velvetDawn.players.getPlayer("player2");

        velvetDawn.game.turns.ready(p1);
        assertTrue(p1.ready);
        assertFalse(p2.ready);

        velvetDawn.game.turns.ready(p2);
        assertTrue(p1.ready);
        assertTrue(p2.ready);

        velvetDawn.game.turns.unready(p1);
        assertFalse(p1.ready);
        assertTrue(p2.ready);

        velvetDawn.game.turns.unready(p2);
        assertFalse(p1.ready);
        assertFalse(p2.ready);
    }

    @Test
    public void test_update_turn_start_time() throws Exception {
        var velvetDawn = this.prepareGame();
        var currentTurnStart = velvetDawn.game.turns.currentTurnData().getTurnStart();

        Thread.sleep(10);

        var updatedTime = velvetDawn.game.turns.updateTurnStartTime();
        assertEquals(
                Math.round(updatedTime),
                Math.round(velvetDawn.game.turns.currentTurnData().getTurnStart()));

        assertNotEquals(
                currentTurnStart.longValue(),
                velvetDawn.game.turns.currentTurnData().getTurnStart().longValue());
    }

    @Test
    public void test_current_turn_time() throws Exception {
        var config = this.getConfig();
        var velvetDawn = new VelvetDawn(config);

        config.turnTime = 10;
        assertEquals(10, velvetDawn.game.turns.currentTurnTime());

        config.turnTime = 500;
        assertEquals(500, velvetDawn.game.turns.currentTurnTime());

        // Test setup is always 5 mins
        config.setupTime = 5;
        velvetDawn.game.phase = Phase.Setup;
        assertEquals(5, velvetDawn.game.turns.currentTurnTime());
    }
}
